native/dlaln2.go

```diff
@@ -0,0 +1,386 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "math"
+
+// Dlaln2 solves a 1×1 or 2×2 linear system of equations of the form
+//  (ca A   - w D) X = scale B,  if trans == false,
+//  (ca A^T - w D) X = scale B,  if trans == true,
+// where A is a na×na real matrix, ca is a real scalar, D is a na×na diagonal
+// real matrix, w is a scalar, real if nw == 1, complex if nw == 2, and X and B
+// are na×1 matrices, real if w is real, complex if w is complex.
+//
+// If w is complex, X and B are represented as na×2 matrices, the first column
+// of each being the real part and the second being the imaginary part.
+//
+// na and nw must be 1 or 2, otherwise Dlaln2 will panic.
+//
+// d1 and d2 are the diagonal elements of D. d2 is not used if na == 1.
+//
+// wr and wi represent the real and imaginary part, respectively, of the scalar
+// w. wi is not used if nw == 1.
+//
+// smin is the desired lower bound on the singular values of A. This should be
+// a safe distance away from underflow or overflow, say, between
+// (underflow/machine precision) and (overflow*machine precision).
+//
+// If both singular values of (ca A - w D) are less than smin, smin*identity
+// will be used instead of (ca A - w D). If only one singular value is less than
+// smin, one element of (ca A - w D) will be perturbed enough to make the
+// smallest singular value roughly smin. If both singular values are at least
+// smin, (ca A - w D) will not be perturbed. In any case, the perturbation will
+// be at most some small multiple of max(smin, ulp*norm(ca A - w D)). The
+// singular values are computed by infinity-norm approximations, and thus will
+// only be correct to a factor of 2 or so.
+//
+// All input quantities are assumed to be smaller than overflow by a reasonable
+// factor.
+//
+// scale is a scaling factor less than or equal to 1 which is chosen so that X
+// can be computed without overflow. X is further scaled if necessary to assure
+// that norm(ca A - w D)*norm(X) is less than overflow.
+//
+// xnorm contains the infinity-norm of X when X is regarded as a na×nw real
+// matrix.
+//
+// ok will be false if (ca A - w D) had to be perturbed to make its smallest
+// singular value greater than smin, otherwise ok will be true.
+//
+// Dlaln2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlaln2(trans bool, na, nw int, smin, ca float64, a []float64, lda int, d1, d2 float64, b []float64, ldb int, wr, wi float64, x []float64, ldx int) (scale, xnorm float64, ok bool) {
+	if na != 1 && na != 2 {
```

Comment:
I'm not sure. `na` is the problem size, so making it a typed constant is not worth it I think, even though it can take only two values. Making `nw` a bool would be natural but again I'm not sure if it's worth it.

Splitting the function into two and using a complex type for w, b, x, is an interesting idea. Similar Go-ification would be applicable also to other routines. It touches on the question of how much we want to follow LAPACK API in such detail. With these internal routines we probably don't need to restict ourselves too much.

I'll add a TODO.